File: julia/src/session.jl
# --------------------------------------------------------------------------------------------
#  Copyright (c) Microsoft Corporation. All rights reserved.
# --------------------------------------------------------------------------------------------

# CopilotSession -- represents a single conversation session with the Copilot CLI.

"""
    CopilotSession

A live conversation session backed by a JSON-RPC connection to the Copilot CLI.

Use `CopilotClient.create_session` to obtain an instance.  The session supports
event subscriptions via `on`, sending messages with `send` / `send_and_wait`,
and graceful teardown with `disconnect`.
"""
mutable struct CopilotSession
    session_id::String
    rpc::JsonRpcClient
    config::SessionConfig
    listeners::Vector{Function}
    tools::Dict{String, Tool}
    connected::Bool
    listener_lock::ReentrantLock
end

function CopilotSession(session_id::String, rpc::JsonRpcClient, config::SessionConfig)
    tools = Dict{String, Tool}()
    for t in config.tools
        if t isa Tool
            tools[t.name] = t
        end
    end
    CopilotSession(session_id, rpc, config, Function[], tools, true, ReentrantLock())
end

# -- Event handling ---------------------------------------------------------------

"""
    on(session, handler) -> Nothing

Register `handler(event::SessionEvent)` for all session events.
"""
function on(session::CopilotSession, handler::Function)
    lock(session.listener_lock) do
        push!(session.listeners, handler)
    end
    return nothing
end

"""
    on(handler, session) -> Nothing

Do-block form: `on(session) do event ... end`.
"""
on(handler::Function, session::CopilotSession) = on(session, handler)

"""Fire an event to all registered listeners."""
function _emit(session::CopilotSession, event::SessionEvent)
    handlers = lock(session.listener_lock) do
        copy(session.listeners)
    end
    for h in handlers
        try
            h(event)
        catch e
            @error "Session event handler error" event_type=event.type exception=(e, catch_backtrace())
        end
    end

    # Also call the config-level handler if present
    if session.config.on_event !== nothing
        try
            session.config.on_event(event)
        catch e
            @error "Config on_event handler error" exception=(e, catch_backtrace())
        end
    end
end

# -- Sending messages -------------------------------------------------------------

"""
    send(session, prompt::String; kwargs...)
    send(session, opts::MessageOptions)

Send a message to the session. Returns immediately; listen for events via `on`.
"""
function send(session::CopilotSession, opts::MessageOptions)
    session.connected || throw(ErrorException("Session is disconnected"))
    params = Dict{String, Any}(
        "sessionId" => session.session_id,
        "prompt"    => opts.prompt,
    )
    if opts.attachments !== nothing
        params["attachments"] = opts.attachments
    end
    if opts.mode !== nothing
        params["mode"] = opts.mode
    end
    send_request(session.rpc, "session/send", params)
    return nothing
end

send(session::CopilotSession, prompt::String; kwargs...) =
    send(session, MessageOptions(; prompt, kwargs...))

"""
    send_and_wait(session, opts; timeout=120) -> SessionEvent

Send a message and block until the session becomes idle, returning the final
`session.idle` event. Raises on timeout.
"""
function send_and_wait(session::CopilotSession, opts::MessageOptions; timeout::Real=120)
    ch = Channel{SessionEvent}(1)
    on(session) do event
        if event.type == "session.idle"
            try put!(ch, event) catch; end
        end
    end

    send(session, opts)

    result = timedwait(timeout) do
        isready(ch)
    end
    if result === :timed_out
        throw(ErrorException("send_and_wait timed out after $(timeout)s"))
    end
    return take!(ch)
end

send_and_wait(session::CopilotSession, prompt::String; kwargs...) =
    send_and_wait(session, MessageOptions(; prompt); kwargs...)

# -- Tool handling ----------------------------------------------------------------

"""Look up and invoke a tool handler registered in this session."""
function _handle_tool_call(session::CopilotSession, params::Dict{String, Any})
    tool_name = get(params, "toolName", get(params, "name", ""))
    tool = get(session.tools, tool_name, nothing)
    if tool === nothing
        return Dict{String, Any}(
            "resultType"       => "failure",
            "textResultForLlm" => "Unknown tool: $tool_name",
        )
    end

    invocation = ToolInvocation(;
        session_id   = session.session_id,
        tool_call_id = get(params, "toolCallId", ""),
        tool_name    = tool_name,
        arguments    = get(params, "arguments", nothing),
    )

    try
        result = tool.handler(invocation)
        if !(result isa ToolResult)
            result = ToolResult(text_result_for_llm=string(result))
        end
        return Dict{String, Any}(
            "resultType"       => TOOL_RESULT_STRINGS[result.result_type],
            "textResultForLlm" => result.text_result_for_llm,
        )
    catch e
        return Dict{String, Any}(
            "resultType"       => "failure",
            "textResultForLlm" => "Tool error: $(sprint(showerror, e))",
        )
    end
end

# -- Permission handling ----------------------------------------------------------

function _handle_permission(session::CopilotSession, params::Dict{String, Any})
    handler = session.config.on_permission_request
    if handler === nothing
        return Dict{String, Any}("decision" => "deny")
    end

    req = PermissionRequest(params)
    try
        decision = handler(req)
        if decision isa Bool
            decision = decision ? "allow" : "deny"
        end
        return Dict{String, Any}("decision" => string(decision))
    catch e
        @error "Permission handler error" exception=(e, catch_backtrace())
        return Dict{String, Any}("decision" => "deny")
    end
end

# -- Lifecycle --------------------------------------------------------------------

"""
    disconnect(session)

Disconnect from the session and release resources. Safe to call multiple times.
"""
function disconnect(session::CopilotSession)
    session.connected || return nothing
    session.connected = false
    try
        send_request(session.rpc, "session/disconnect",
            Dict{String, Any}("sessionId" => session.session_id); timeout=5)
    catch; end
    return nothing
end

File: zig/src/session.zig
// session.zig -- CopilotSession wraps a single conversation session.
//
// A session is created by the client via `session.create` / `session.resume`
// and provides send, event subscription, tool registration, and lifecycle
// management.

const std = @import("std");
const types = @import("types.zig");
const tools_mod = @import("tools.zig");
const jsonrpc = @import("jsonrpc.zig");

const Allocator = std.mem.Allocator;
const JsonValue = types.JsonValue;
const SdkError = types.SdkError;
const SessionEvent = types.SessionEvent;
const MessageOptions = types.MessageOptions;
const ToolInvocation = types.ToolInvocation;
const ToolResult = types.ToolResult;
const PermissionRequest = types.PermissionRequest;
const PermissionResult = types.PermissionResult;
const UserInputRequest = types.UserInputRequest;
const UserInputResponse = types.UserInputResponse;
const ToolRegistry = tools_mod.ToolRegistry;
const ToolDefinition = types.ToolDefinition;
const ToolHandlerFn = tools_mod.ToolHandlerFn;

// ---------------------------------------------------------------------------
// Callback types
// ---------------------------------------------------------------------------

pub const EventCallback = *const fn (event: SessionEvent) void;
pub const PermissionHandlerFn = *const fn (req: PermissionRequest, session_id: []const u8) anyerror!PermissionResult;
pub const UserInputHandlerFn = *const fn (req: UserInputRequest, session_id: []const u8) anyerror!UserInputResponse;
pub const ExitPlanModeHandlerFn = *const fn (req: types.ExitPlanModeRequest, session_id: []const u8) anyerror!types.ExitPlanModeResponse;

// ---------------------------------------------------------------------------
// EventSubscription
// ---------------------------------------------------------------------------

pub const EventSubscription = struct {
    id: u64,
    session: *CopilotSession,

    /// Remove this subscription.
    pub fn unsubscribe(self: EventSubscription) void {
        self.session.removeSubscription(self.id);
    }
};

// ---------------------------------------------------------------------------
// CopilotSession
// ---------------------------------------------------------------------------

pub const CopilotSession = struct {
    allocator: Allocator,
    session_id: []const u8,
    workspace_path: ?[]const u8 = null,
    transport: *jsonrpc.JsonRpcTransport,
    tool_registry: ToolRegistry,
    listeners: std.ArrayList(Listener),
    next_listener_id: u64 = 1,
    permission_handler: ?PermissionHandlerFn = null,
    user_input_handler: ?UserInputHandlerFn = null,
    exit_plan_mode_handler: ?ExitPlanModeHandlerFn = null,
    trace_context_provider: ?types.TraceContextProvider = null,
    last_event: ?SessionEvent = null,
    idle: bool = true,

    const Listener = struct {
        id: u64,
        event_type: ?[]const u8, // null means all events
        callback: EventCallback,
    };

    pub fn init(
        allocator: Allocator,
        session_id: []const u8,
        workspace_path: ?[]const u8,
        transport: *jsonrpc.JsonRpcTransport,
        trace_context_provider: ?types.TraceContextProvider,
    ) CopilotSession {
        return .{
            .allocator = allocator,
            .session_id = session_id,
            .workspace_path = workspace_path,
            .transport = transport,
            .trace_context_provider = trace_context_provider,
            .tool_registry = ToolRegistry.init(allocator),
            .listeners = std.ArrayList(Listener).init(allocator),
        };
    }

    pub fn deinit(self: *CopilotSession) void {
        self.tool_registry.deinit();
        self.listeners.deinit();
    }

    // -------------------------------------------------------------------
    // Sending messages
    // -------------------------------------------------------------------

    /// Send a message to the session (non-blocking).
    pub fn send(self: *CopilotSession, opts: MessageOptions) SdkError![]const u8 {
        self.idle = false;
        var obj = std.json.ObjectMap.init(self.allocator);
        defer obj.deinit();

        obj.put("sessionId", .{ .string = self.session_id }) catch return SdkError.AllocationFailed;
        obj.put("prompt", .{ .string = opts.prompt }) catch return SdkError.AllocationFailed;

        const params = JsonValue{ .object = obj };
        const result = try self.transport.sendRequest("session.send", params);

        // Extract messageId from result
        if (result == .object) {
            if (result.object.get("messageId")) |mid| {
                if (mid == .string) return mid.string;
            }
        }
        return "";
    }

    /// Send a message and block until the session becomes idle or times out.
    pub fn sendAndWait(self: *CopilotSession, opts: MessageOptions, timeout_ms: ?u64) SdkError!?SessionEvent {
        _ = try self.send(opts);

        const deadline = timeout_ms orelse 60_000;
        const start = std.time.milliTimestamp();

        while (!self.idle) {
            const elapsed: u64 = @intCast(std.time.milliTimestamp() - start);
            if (elapsed >= deadline) return SdkError.Timeout;
            self.transport.processOneMessage() catch |err| {
                if (err == SdkError.ConnectionClosed) return SdkError.ConnectionClosed;
                // For other transient errors, keep trying
                continue;
            };
        }

        return self.last_event;
    }

    // -------------------------------------------------------------------
    // Event subscriptions
    // -------------------------------------------------------------------

    /// Subscribe to all session events.
    pub fn on(self: *CopilotSession, callback: EventCallback) EventSubscription {
        const id = self.next_listener_id;
        self.next_listener_id += 1;
        self.listeners.append(.{
            .id = id,
            .event_type = null,
            .callback = callback,
        }) catch {};
        return .{ .id = id, .session = self };
    }

    /// Subscribe to events of a specific type (e.g. "assistant.message").
    pub fn onEvent(self: *CopilotSession, event_type: []const u8, callback: EventCallback) EventSubscription {
        const id = self.next_listener_id;
        self.next_listener_id += 1;
        self.listeners.append(.{
            .id = id,
            .event_type = event_type,
            .callback = callback,
        }) catch {};
        return .{ .id = id, .session = self };
    }

    /// Remove a subscription by id.
    pub fn removeSubscription(self: *CopilotSession, sub_id: u64) void {
        var i: usize = 0;
        while (i < self.listeners.items.len) {
            if (self.listeners.items[i].id == sub_id) {
                _ = self.listeners.orderedRemove(i);
                return;
            }
            i += 1;
        }
    }

    /// Deliver an event to all matching listeners.
    pub fn emitEvent(self: *CopilotSession, event: SessionEvent) void {
        self.last_event = event;
        if (event.isIdle()) self.idle = true;

        for (self.listeners.items) |listener| {
            if (listener.event_type) |wanted| {
                if (!std.mem.eql(u8, wanted, event.event_type)) continue;
            }
            listener.callback(event);
        }
    }

    // -------------------------------------------------------------------
    // Tool registration
    // -------------------------------------------------------------------

    /// Register a tool handler for this session.
    pub fn registerTool(self: *CopilotSession, definition: types.ToolDefinition, handler: ToolHandlerFn) !void {
        try self.tool_registry.register(definition, handler);
    }

    /// Handle an incoming tool call from the server.
    pub fn handleToolCall(self: *CopilotSession, invocation: ToolInvocation) anyerror!ToolResult {
        return self.tool_registry.dispatch(invocation);
    }

    // -------------------------------------------------------------------
    // Permission & user input handlers
    // -------------------------------------------------------------------

    pub fn registerPermissionHandler(self: *CopilotSession, handler: PermissionHandlerFn) void {
        self.permission_handler = handler;
    }

    pub fn registerUserInputHandler(self: *CopilotSession, handler: UserInputHandlerFn) void {
        self.user_input_handler = handler;
    }

    pub fn registerExitPlanModeHandler(self: *CopilotSession, handler: ExitPlanModeHandlerFn) void {
        self.exit_plan_mode_handler = handler;
    }

    // -------------------------------------------------------------------
    // Lifecycle
    // -------------------------------------------------------------------

    /// Destroy the session on the server side.
    pub fn destroy(self: *CopilotSession) SdkError!void {
        var obj = std.json.ObjectMap.init(self.allocator);
        defer obj.deinit();
        obj.put("sessionId", .{ .string = self.session_id }) catch return SdkError.AllocationFailed;
        _ = try self.transport.sendRequest("session.destroy", .{ .object = obj });
    }

    /// Abort the current processing turn.
    pub fn abort(self: *CopilotSession) SdkError!void {
        var obj = std.json.ObjectMap.init(self.allocator);
        defer obj.deinit();
        obj.put("sessionId", .{ .string = self.session_id }) catch return SdkError.AllocationFailed;
        _ = try self.transport.sendRequest("session.abort", .{ .object = obj });
    }

    /// Get the list of messages in this session.
    pub fn getMessages(self: *CopilotSession) SdkError!JsonValue {
        var obj = std.json.ObjectMap.init(self.allocator);
        defer obj.deinit();
        obj.put("sessionId", .{ .string = self.session_id }) catch return SdkError.AllocationFailed;
        return try self.transport.sendRequest("session.getMessages", .{ .object = obj });
    }

    /// Get the session ID.
    pub fn getId(self: *const CopilotSession) []const u8 {
        return self.session_id;
    }

    /// Get the workspace path, if set.
    pub fn getWorkspacePath(self: *const CopilotSession) ?[]const u8 {
        return self.workspace_path;
    }
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

fn noopCallback(_: SessionEvent) void {}

test "CopilotSession event subscribe and unsubscribe" {
    const allocator = std.testing.allocator;

    var dummy_r: [1]u8 = undefined;
    var rstream = std.io.fixedBufferStream(&dummy_r);
    var dummy_w: [1]u8 = undefined;
    var wstream = std.io.fixedBufferStream(&dummy_w);

    var transport = jsonrpc.JsonRpcTransport.init(
        allocator,
        wstream.writer().any(),
        rstream.reader().any(),
    );
    defer transport.deinit();

    var session = CopilotSession.init(allocator, "test-session", null, &transport, null);
    defer session.deinit();

    const sub = session.on(noopCallback);
    try std.testing.expectEqual(@as(usize, 1), session.listeners.items.len);
    sub.unsubscribe();
    try std.testing.expectEqual(@as(usize, 0), session.listeners.items.len);
}

test "CopilotSession emitEvent sets idle" {
    const allocator = std.testing.allocator;

    var dummy_r: [1]u8 = undefined;
    var rstream = std.io.fixedBufferStream(&dummy_r);
    var dummy_w: [1]u8 = undefined;
    var wstream = std.io.fixedBufferStream(&dummy_w);

    var transport = jsonrpc.JsonRpcTransport.init(
        allocator,
        wstream.writer().any(),
        rstream.reader().any(),
    );
    defer transport.deinit();

    var session = CopilotSession.init(allocator, "test-session", null, &transport, null);
    defer session.deinit();
    session.idle = false;

    session.emitEvent(.{ .event_type = "session.idle" });
    try std.testing.expect(session.idle);
}

test "CopilotSession getId returns session_id" {
    const allocator = std.testing.allocator;

    var dummy_r: [1]u8 = undefined;
    var rstream = std.io.fixedBufferStream(&dummy_r);
    var dummy_w: [1]u8 = undefined;
    var wstream = std.io.fixedBufferStream(&dummy_w);

    var transport = jsonrpc.JsonRpcTransport.init(
        allocator,
        wstream.writer().any(),
        rstream.reader().any(),
    );
    defer transport.deinit();

    var session = CopilotSession.init(allocator, "abc-123", null, &transport, null);
    defer session.deinit();
    try std.testing.expectEqualStrings("abc-123", session.getId());
}
